lib/improver/utilities/mathematical_operations.py

```diff
@@ -0,0 +1,228 @@
+
+        upper_bounds_cube = (
+             cube.extract(
+                 iris.Constraint(
+                     coord_values={self.coord_name_to_integrate:
+                                   upper_bounds})))
+        lower_bounds_cube = (
+            cube.extract(
+                iris.Constraint(
+                    coord_values={self.coord_name_to_integrate:
+                                  lower_bounds})))
+
+        # Determine which cube to copy in order to have the most appropriate
+        # points within the coordinate that is being integrated.
+        if self.direction_of_integration == "positive":
+            integrated_cube = upper_bounds_cube.copy()
+        elif self.direction_of_integration == "negative":
+            integrated_cube = lower_bounds_cube.copy()
+        integrated_cube.data = np.zeros(lower_bounds_cube.shape)
+
+        # Create a zip for looping over.
+        levels_tuple = zip(
+            upper_bounds_cube.slices_over(self.coord_name_to_integrate),
+            lower_bounds_cube.slices_over(self.coord_name_to_integrate),
+            integrated_cube.slices_over(self.coord_name_to_integrate))
+
+        # Perform the integration
+        stride_sum = 0
+        integrated_cubelist = iris.cube.CubeList([])
+        for (upper_bounds_slice, lower_bounds_slice,
+             integrated_slice) in levels_tuple:
+            upper_bound = (
+                upper_bounds_slice.coord(
+                    self.coord_name_to_integrate).points.item())
+            lower_bound = (
+                lower_bounds_slice.coord(
+                    self.coord_name_to_integrate).points.item())
+            if not self.start_point and not self.end_point:
+                pass
+            elif self.start_point:
+                if self.direction_of_integration == "positive":
+                    if lower_bound <= self.start_point:
+                        continue
+                elif self.direction_of_integration == "negative":
+                    if upper_bound >= self.start_point:
+                        continue
+            elif self.end_point:
+                if self.direction_of_integration == "positive":
+                    if upper_bound >= self.end_point:
+                        continue
+                elif self.direction_of_integration == "negative":
+                    if lower_bound <= self.end_point:
+                        continue
+            stride = np.abs(upper_bound - lower_bound)
+            upper_half_of_stride = upper_bounds_slice.data * 0.5 * stride
+            lower_half_of_stride = lower_bounds_slice.data * 0.5 * stride
+            stride_sum += lower_half_of_stride + upper_half_of_stride
+            integrated_slice.data = stride_sum
+            integrated_cubelist.append(integrated_slice.copy())
+
+        # Merge resulting cubes back together
+        integrated_cube = integrated_cubelist.merge_cube()
```

Comment:
Lines 173-214 could be refactored into a function "perform_integration" that updates self.integrated_cube.

Reply:
I'll deal with this in a separate commit.

Reply:
Now refactored as suggested.